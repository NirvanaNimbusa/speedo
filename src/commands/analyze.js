import ora from 'ora'
import yargs from 'yargs'
import SauceLabs from 'saucelabs'

import { getMetricParams } from '../utils'
import { ERROR_MISSING_CREDENTIALS, ANALYZE_CLI_PARAMS } from '../constants'

export const command = 'analyze [params...] <build>'
export const desc = 'Analyze results of prerun performance tests.'
export const builder = ANALYZE_CLI_PARAMS

export const handler = async (argv) => {
    const username = process.env.SAUCE_USERNAME || argv.user
    const accessKey = process.env.SAUCE_ACCESS_KEY || argv.key
    const metrics = getMetricParams(argv)

    /**
     * check if username and access key are available
     */
    if (!username || !accessKey) {
        yargs.showHelp()
        // eslint-disable-next-line no-console
        console.error(ERROR_MISSING_CREDENTIALS)
        return process.exit(1)
    }

    /**
     * check if job already exists
     */
    const user = new SauceLabs(username, accessKey)

    /**
     * find build
     */
    let builds = null
    const status = ora(`Fetch last builds from ${username}`).start()
    try {
        builds = await user.listBuilds(
            username,
            // fetch the last 5 builds and hope the desired one is in there
            { limit: 5, subaccounts: false }
        )
    } catch (e) {
        status.fail(`Couldn't fetch builds: ${e.stack}`)
        return process.exit(1)
    }

    const build = builds.find((build) => build.name === argv.build)
    if (!build) {
        status.fail(`Couldn't find build with name "${argv.build}"`)
        return process.exit(1)
    }

    status.succeed()
    status.start(`Fetch jobs from build "${argv.build}"`)
    let buildJobs = null
    try {
        buildJobs = (await user.listBuildJobs(build.id, { full: true })).jobs
    } catch (e) {
        status.fail(`Couldn't fetch job from build with name "${argv.build}": ${e.stack}`)
        return process.exit(1)
    }

    const specificJobs = buildJobs.filter((job) => argv.name && job.name === argv.name)
    if (argv.name && specificJobs.length) {
        buildJobs = specificJobs
    }

    status.succeed()
    status.start(`Analyze performance of ${buildJobs.length} jobs`)
    let performanceResults = null
    try {
        for (const job of buildJobs) {
            const perfTests = await user.assertPerformanceRegression(job.id, {
                metricNames: metrics
            })
            console.log(perfTests)
        }
    } catch (e) {
        status.fail(`Couldn't fetch performance results: ${e.stack}`)
        return process.exit(1)
    }

    status.succeed()
    // console.log(JSON.stringify(results, null, 4))

}
