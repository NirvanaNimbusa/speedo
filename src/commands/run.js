import tmp from 'tmp'
import ora from 'ora'
import path from 'path'
import yargs from 'yargs'
import ordinal from 'ordinal'
import SauceLabs from 'saucelabs'

import runPerformanceTest from '../runner'
import { printResult, waitFor, getMetricParams, getJobUrl } from '../utils'
import { ERROR_MISSING_CREDENTIALS, REQUIRED_TESTS_FOR_BASELINE_COUNT, RUN_CLI_PARAMS } from '../constants'

export const command = 'run [params...] <site>'
export const desc = 'Run performance tests on any website.'
export const builder = RUN_CLI_PARAMS

export const handler = async (argv) => {
    const username = process.env.SAUCE_USERNAME || argv.user
    const accessKey = process.env.SAUCE_ACCESS_KEY || argv.key
    const jobName = argv.name || `Performance test for ${argv.site}`
    const buildName = argv.build || `${jobName} - ${(new Date()).toString()}`
    const metrics = getMetricParams(argv)

    /**
     * check if username and access key are available
     */
    if (!username || !accessKey) {
        yargs.showHelp()
        // eslint-disable-next-line no-console
        console.error(ERROR_MISSING_CREDENTIALS)
        return process.exit(1)
    }

    const status = ora(`Start performance test run with user ${username} on page ${argv.site}...`).start()

    const logDir = argv.logDir
        ? path.resolve(process.cwd(), argv.logDir)
        : tmp.dirSync().name

    /**
     * check if job already exists
     */
    const user = new SauceLabs({
        user: username,
        key: accessKey,
        region: argv.region
    })

    /**
     * find if job already exists
     */
    let job = null
    try {
        job = await user.listJobs(
            username,
            { name: jobName, limit: 10 }
        )
        status.succeed()
    } catch (e) {
        status.fail(`Couldn't fetch job: ${e.stack}`)
        return process.exit(1)
    }

    /**
     * create baseline if not enough tests have been executed
     */
    if (job.jobs.length < REQUIRED_TESTS_FOR_BASELINE_COUNT) {
        status.start(`Couldn't find baseline for job with name ${jobName}, creating baseline...`)

        const testCnt = REQUIRED_TESTS_FOR_BASELINE_COUNT - job.jobs.length
        await Promise.all([...Array(testCnt)].map(
            () => runPerformanceTest(username, accessKey, argv, jobName, undefined, logDir)))
        status.succeed()
    }

    /**
     * run single test
     */
    status.start('Run performance test...')
    let { result, sessionId } = await runPerformanceTest(
        username, accessKey, argv, jobName, buildName, logDir)

    /**
     * retry performance test
     */
    const retriedJobs = []
    if (result.result !== 'pass' && argv.retry) {
        for (let retry = 1; retry <= argv.retry; ++retry) {
            retriedJobs.push(sessionId)
            status.text = `Run performance test (${ordinal(retry)} retry)...`

            const retriedResult = await runPerformanceTest(
                username, accessKey, argv, jobName, buildName, logDir)

            result = retriedResult.result
            sessionId = retriedResult.sessionId

            /**
             * continue command if job has finally passed
             */
            if (result.result === 'pass') {
                break
            }
        }
    }

    status.succeed()

    /**
     * wait until job completes
     */
    try {
        status.start('Wait for job to finish...')
        await waitFor(
            () => user.getJob(username, sessionId),
            (jobDetails) => jobDetails.status === 'complete'
        )
        status.succeed()
    } catch (e) {
        status.fail(e.message)
        return process.exit(1)
    }

    /**
     * download performance logs
     */
    let performanceLog
    try {
        status.start('Download performance logs...')
        performanceLog = await user.downloadJobAsset(
            sessionId,
            'performance.json',
            path.join(logDir, 'performance.json'))
        status.succeed()
    } catch (e) {
        status.fail(`Couldn't download performance results due to: ${e.stack}`)
        return process.exit(1)
    }

    /**
     * download trace file if requested
     */
    if (argv.traceLogs) {
        status.start('Download trace logs...')

        const loaderId = performanceLog[0].loaderId
        await user.downloadJobAsset(
            sessionId,
            `_tracelog_${loaderId}.json.gz`,
            path.join(logDir, 'trace.json'))
    }

    /**
     * update job if performance check fails
     */
    try {
        status.start('Updating job status...')
        await user.updateJob(username, sessionId, { passed: result.result === 'pass' })
        status.succeed()
    } catch (e) {
        status.fail(`Couldn't update job due to: ${e.stack}`)
        status.stopAndPersist({ text: 'continuing ...' })
    }

    status.stopAndPersist({
        text: `Stored performance logs in ${logDir}`,
        symbol: 'ðŸ“ƒ'
    })

    printResult(result, performanceLog[0], metrics)

    status.stopAndPersist({
        text: `Check out job at ${getJobUrl(argv, sessionId)}`,
        symbol: 'ðŸ‘€'
    })

    /**
     * displayed retried jobs that failed
     */
    if (retriedJobs.length) {
        console.log( // eslint-disable-line no-console
            '\nFailed Performance Tests that were rerun:\n' +
            retriedJobs.map(id => getJobUrl(argv, id)).join('\n')
        )
    }

    process.exit(result.result === 'pass' ? 0 : 1)
}
