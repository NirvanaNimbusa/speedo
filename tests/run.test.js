import yargs from 'yargs'

import { handler } from '../src/commands/run'

const processExit = ::process.exit

beforeEach(() => {
    process.exit = jest.fn()
    delete process.env.SAUCE_USERNAME
    delete process.env.SAUCE_ACCESS_KEY
})

test('run should fail if no auth is provided', async () => {
    await handler({})
    expect(process.exit).toBeCalledWith(1)
    expect(yargs.showHelp).toBeCalledTimes(1)
})

test('run', async () => {
    await handler({ user: 'foo', key: 'bar', metric: ['load', 'speedIndex'] })
})

afterEach(() => {
    process.exit = processExit
})
