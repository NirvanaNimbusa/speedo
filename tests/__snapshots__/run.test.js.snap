// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should run successfully 1`] = `
Array [
  Array [],
  Array [
    "Run performance test...",
  ],
  Array [
    "Wait for job to finish...",
  ],
  Array [
    "Download performance logs...",
  ],
  Array [
    "Updating job status...",
  ],
]
`;

exports[`should run successfully 2`] = `
Array [
  Array [
    Object {
      "symbol": "ðŸ“ƒ",
      "text": "Stored performance logs in /some/tmpDir",
    },
  ],
  Array [
    Object {
      "symbol": "ðŸ‘€",
      "text": "Check out job at https://saucelabs.com/tests/foobar",
    },
  ],
]
`;

exports[`should store tracelogs if path provided 1`] = `
Array [
  Array [
    "foobar123",
    "_tracelog_3b879028728e33d3e1e7789128c5ad6ad1371278.json.gz",
    "/foo/bar/trace.json",
  ],
]
`;
